src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs

```diff
@@ -0,0 +1,87 @@
+        private readonly RenderingEngineOptions _options;
+        private readonly IGraphQLClientFactory _graphQLClientFactory;
+        private readonly IGraphQLRequestBuilder _graphQLRequestBuilder;
+        private readonly IConfiguration _configuration;
+
+        public CustomRenderingEngineMiddleware(RequestDelegate next, ISitecoreLayoutClient layoutService, ISitecoreLayoutRequestMapper requestMapper, IOptions<RenderingEngineOptions> options, IGraphQLClientFactory graphQLClientFactory, IGraphQLRequestBuilder graphQLRequestBuilder, IConfiguration configuration)
+        {
+            _next = next;
+            _requestMapper = requestMapper;
+            _options = options.Value;
+            _graphQLClientFactory = graphQLClientFactory;
+            _graphQLRequestBuilder = graphQLRequestBuilder;
+            _layoutService = layoutService;
+            _configuration= configuration;
+        }              
+
+        public async Task Invoke(HttpContext httpContext, IViewComponentHelper viewComponentHelper, IHtmlHelper htmlHelper)
+        {
+            SitecoreLayoutRequest sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+
+            if (httpContext.GetSitecoreRenderingContext() == null)
+            {
+                SitecoreLayoutResponse response = await GetSitecoreLayoutResponse(httpContext).ConfigureAwait(continueOnCapturedContext: false);
+
+                //Check not found page and set status code - start
+                if (response.Request.Path() == _configuration.GetSection(MvpSiteSettings.Key).Get<MvpSiteSettings>().NotFoundPage) { httpContext.Response.StatusCode = (int)HttpStatusCode.NotFound; }
+                //Check not found page and set status code - end
+
+                SitecoreRenderingContext renderingContext = new()
+                {
+                    Response = response,
+                    RenderingHelpers = new RenderingHelpers(viewComponentHelper, htmlHelper)
+                };
+                httpContext.SetSitecoreRenderingContext(renderingContext);
+            }
+            else
+            {
+                httpContext.GetSitecoreRenderingContext().RenderingHelpers = new RenderingHelpers(viewComponentHelper, htmlHelper);
+            }
+
+            foreach (Action<HttpContext> postRenderingAction in (IEnumerable<Action<HttpContext>>)_options.PostRenderingActions)
+                postRenderingAction(httpContext);
+
+            httpContext.Items.Add((object)nameof(RenderingEngineMiddleware), (object)null);
+
+            await _next(httpContext).ConfigureAwait(continueOnCapturedContext: false);
+        }
+
+        private async Task<SitecoreLayoutResponse> GetSitecoreLayoutResponse(HttpContext httpContext)
+        {
+            //intercept for not found logic - start
+            CustomGraphQlLayoutServiceHandler customGraphQlLayoutServiceHandler = new(_configuration,_graphQLRequestBuilder, _graphQLClientFactory);
```

Comment:
Don't use `new` to create instances like this, it should be injected into the class ctor via DI